### PR TITLE
KD & Klient metrics

### PR DESCRIPTION
**Pull request by cihangir**

### CODE REVIEW

**Comment by cihangir** [comment]
i forgot @rjeczalik  was off today assigning to @ppknap

File: go/src/koding/kites/kloud/main.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func main() {
 	if err != nil {
 		log.Fatal(err)
 	}
+	defer k.Close()
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/main.go]: Just a note.

If statsd package does a cleanup in its Close method (like flush events etc.), it may not work when kloud is stopped / restarted by systemd - once go program receives a signal it executes a os.Exit(0), which do not run deferred methods.

File: go/src/koding/kites/kloud/metrics/metrics.go [hunk 2]
```diff
@@ -55,9 +56,17 @@ func (p *Publisher) Publish(r *kite.Request) (interface{}, error) {
 
 	for _, data := range req.Data {
 		_, err := p.conn.Write(data)
-		if err != nil {
-			return nil, err
+		if err == nil {
```
Comment (rjeczalik) [inline go/src/koding/kites/kloud/metrics/metrics.go]: This branch can be removed, L69 already covers that.

File: go/src/koding/kites/metrics/cron.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/koding/logging"
+)
+
+type kloudKite interface {
+	Call(method string, req, resp interface{}) error
+}
+
+// StartCron starts klient metrics publisher.
+func StartCron(kloud kloudKite, log logging.Logger) {
+	for range time.NewTicker(time.Second * 3).C {
```
Comment (rjeczalik) [inline go/src/koding/kites/metrics/cron.go]: `time.NewTicker(...).C` -> `time.Tick(...)`

File: go/src/koding/kites/metrics/cron.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+func StartCronWithMetrics(kloud kloudKite, log logging.Logger, m *Metrics) {
+	for range time.NewTicker(time.Second * 3).C {
+		processWithMetrics(kloud, log, m)
+	}
+}
+
+func process(kloud kloudKite, log logging.Logger) {
+	m, err := New("kd")
+	if err != nil {
+		log.Error(err.Error())
```
Comment (rjeczalik) [inline go/src/koding/kites/metrics/cron.go]: log.Error("%s", err)

If err.Error() contains `%` characters, the log line will be misformatted.

File: go/src/koding/kites/metrics/cron.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		}
+
+		log.Debug("publishing %d elements", len(res))
+
+		req := &metrics.PublishRequest{
+			Data: metrics.GzippedPayload(res),
+		}
+
+		var resp interface{}
+		if err := kloud.Call("metrics.publish", req, resp); err != nil {
```
Comment (rjeczalik) [inline go/src/koding/kites/metrics/cron.go]: Can be just `kloud.Call("metrics.publish", req, nil)`

File: go/src/koding/kites/metrics/cron.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+			log.Error(err.Error())
```
Comment (rjeczalik) [inline go/src/koding/kites/metrics/cron.go]: ditto

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/kites/kloud/kloud/kloud.go [hunk 7]
```diff
@@ -389,6 +405,24 @@ func New(conf *Config) (*Kloud, error) {
...
+// HandleFunc adds our middlewares into kite handlers.
+func (k *Kloud) HandleFunc(pattern string, f kite.HandlerFunc) *kite.Method {
+	f = kitemetrics.WrapKiteHandler(k.Stats, pattern, f)
+	return k.Kite.HandleFunc(pattern, f)
+}
+
+// Close closes the underlying connections.
+func (k *Kloud) Close() error {
+	if k.metricsProxy != nil {
+		return k.metricsProxy.Close()
```
Comment (ppknap) [inline go/src/koding/kites/kloud/kloud/kloud.go]: Shouldn't we close `k.Kite` and `k.Stats` as well here (rm `return` keyword)?

File: go/src/koding/kites/metrics/wrappers.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		if err != nil {
+			msg := err.Error()
+			if len(msg) > 20 {
+				msg = msg[:20]
+			}
+			tags = AppendTag(tags, "err_message", msg)
+		}
+
+		dd.Count(metricName+"_call_count", 1, tags, 1)
+		dd.Timing(metricName+"_timing", dur, tags, 1)
```
Comment (ppknap) [inline go/src/koding/kites/metrics/wrappers.go]: How much will these calls affect cli times?

**Review by ppknap — APPROVED** [review]

File: go/src/koding/kites/kloud/main.go (hunk 1 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/main.go]: It doesnt not do any flush or smth but i just wanted to close/free resources immediately. Thanks for the note. These come default in socialapi runner so didn't give a second thought about whether if kloud supports graceful shutdown or not.

File: go/src/koding/kites/kloud/metrics/metrics.go (hunk 2 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/metrics/metrics.go]: if error is nil, no need to do a type casting. This check is in a critical path and most of the time it will be nil (in a healthy environment)

File: go/src/koding/kites/kloud/main.go (hunk 1 above)
Comment (rjeczalik) [inline go/src/koding/kites/kloud/main.go]: I think socialapi suffers from the same problem, all the defers here won't get executed e.g. on sigterm:

https://github.com/koding/koding/blob/master/go/src/socialapi/workers/api/main.go#L38

File: go/src/koding/kites/metrics/cron.go (hunk 3 above)
Comment (cihangir) [inline go/src/koding/kites/metrics/cron.go]: 👍  i wasplanningg to close the channel, hence the usage

File: go/src/koding/kites/metrics/wrappers.go (hunk 8 above)
Comment (cihangir) [inline go/src/koding/kites/metrics/wrappers.go]: adds <0.4 ms to each command
overall whole metrics functionality adds <1.5 ms to each command

will send the benchmarks in another PR

File: go/src/koding/kites/kloud/main.go (hunk 1 above)
Comment (cihangir) [inline go/src/koding/kites/kloud/main.go]: nope it doesnt see: 
https://github.com/koding/koding/blob/master/go/src/socialapi/workers/api/main.go#L102
and
https://github.com/koding/koding/blob/master/go/src/vendor/github.com/koding/runner/runner.go#L214-L233

Comment (rjeczalik) [inline go/src/koding/kites/kloud/main.go]: Oh, it's in a separate project, I missed that, sorry.
